romancal/datamodels/tests/test_models.py

```diff
@@ -1,16 +1,186 @@
+        sub_dict = meta_dict
+        for subkey in keylist[:-1]:
+            sub_dict = sub_dict[subkey]
+        del sub_dict[keylist[-1]]
+        return sub_dict
+
+    # Cycle through each key/value pair in meta dictionary
+    for keypathlist in getkeypaths(meta):
+        keypathname = '.'.join(keypathlist)
+
+        # Create temporary copy of meta dictionary and remove one key/value pair
+        badmeta = deepcopy(meta)
+        delkeypath(badmeta,keypathname)
+
+        # Save asdf file with badmeta dictionary as metadata tree
+        with asdf.AsdfFile(badmeta) as af:
+            af.write_to(file_path)
+
+        # Test for warning that required keyword is missing
+        with DataModels.open(file_path) as model:
+            with pytest.warns(ValidationWarning) as record:
+                model.validate()
+        assert (len(record) > 0)
+
+# Testing core schema
+def test_core_schemas(tmp_path):
+
+    # Define meta dictionary of required key / value pairs
+    meta = {"meta":
+                {"instrument": {
+                     "detector": "WFI01",
+                     "optical_element": "F158",
+                     "name": "WFI"
+                 },
+                 "telescope": "ROMAN"
+                 }
+            }
+
+    # Test schema for required key/value pairs
+    confirm_required_keywords(tmp_path, meta)
+
+# Testing all reference file schemas
+def test_referencefile_model_schemas(tmp_path):
+
+    # Define base meta dictionary of required key / value pairs for reference files
+    # NOTE 1: date is commented out because it is not properly failing its test
+    #         Ticket: https://github.com/spacetelescope/stdatamodels/issues/23
+    # NOTE 2: core.schema key/value pairs commented out due to not properly failing their tests
+    #         Ticket: https://github.com/spacetelescope/stdatamodels/issues/7
+
+    meta = {"meta":
+                {
+                    "author": "Space Telescope Science Institute",
+                    "description": "Test file",
+                    "pedigree": "Test",
+                    "reftype": "REFERENCEFILE",
+                    "useafter": Time('1999-01-01T00:00:00.123456789',format='isot', scale='utc'),
+#                    "date": Time('1999-01-01T00:00:00.123456789',format='isot', scale='utc'),
+#                     # Key/value pairs from core.schema
+#                     "instrument": {
+#                         "detector": "WFI01",
+#                         "optical_element": "F158",
+#                         "name": "WFI"
+#                     },
+#                     "telescope": "ROMAN",
+                }
+    }
+
+    # Set temporary asdf file
+    file_path = tmp_path / "test.asdf"
+
+    # Cycle through reference file models to test each
+    for model_class in iter_subclasses(DataModels.referencefile.ReferenceFileModel):
+        # Test base referencefile model
+        if (model_class == DataModels.referencefile.ReferenceFileModel):
```

Comment:
I think this `if` statement is a sign that we need separate tests for `ReferenceFileModel` and actual reference models like `FlatModel`.  It's also nice to parametrize the test so that we can easily see what models are failing, otherwise the test will just halt at the first failure.  I'd recommend we set up like this:

```python
def test_reference_file_model_base(tmp_path):
  # ReferenceFileModel tests


@pytest.mark.parametrize("model_class", list(iter_subclasses(ReferenceFileModel)))
def test_reference_file_model(tmp_path, model_class):
  # Common tests for all ReferenceFileModel subclasses


def test_flat_model(tmp_path):
  # FlatModel tests
```

We'd need to add a flag to iter_subclasses that prevents it from yielding the original parent class.

Reply:
Refactored.